velox/dwio/common/BitPackDecoder.h

```diff
@@ -781,6 +337,116 @@ inline void unpack<uint32_t>(
     default:
       VELOX_UNREACHABLE("invalid bitWidth");
   }
+}
+
+/// Unpack numValues number of input values from inputBuffer. The results
+/// will be written to result. numValues must be a multiple of 8. The
+/// caller needs to make sure the inputBufferLen contains at least numValues
+/// number of packed values. The inputBits and result pointers will be updated
+/// to the next to read position after this call.
+template <typename T>
+inline void unpack(
```

Comment:
Can we keep only this interface and move the rest into cpp files?  I don't think inlining at this level is very important to performance (let me know if you find it affect performance in the benchmark)

Reply:
Based on the benchmark results, putting the functions in the .cpp files does affect the performance, as you can see the results of output `uint8_t` in the below table, both `AVX2` path and `AVX-512` path will be affected. 
```
============================================================================
[...]/tests/BitPackDecoderBenchmarknew.cpp     relative  time/iter   iters/s
============================================================================
velox_unpack_fullrows_switch_1_8                            2.79ms    357.87
avx512_new_unpack_fullrows_switch_1_8           694.63%   402.27us     2.49K
----------------------------------------------------------------------------
velox_unpack_fullrows_switch_2_8                            2.79ms    358.75
avx512_new_unpack_fullrows_switch_2_8            404.7%   688.76us     1.45K
----------------------------------------------------------------------------
velox_unpack_fullrows_switch_3_8                            2.79ms    358.68
avx512_new_unpack_fullrows_switch_3_8            389.9%   715.07us     1.40K
----------------------------------------------------------------------------
velox_unpack_fullrows_switch_4_8                            2.78ms    359.41
avx512_new_unpack_fullrows_switch_4_8           391.74%   710.24us     1.41K
----------------------------------------------------------------------------
velox_unpack_fullrows_switch_5_8                            2.78ms    359.27
avx512_new_unpack_fullrows_switch_5_8           388.25%   716.92us     1.39K
----------------------------------------------------------------------------
velox_unpack_fullrows_switch_6_8                            2.78ms    359.77
avx512_new_unpack_fullrows_switch_6_8           387.16%   717.93us     1.39K
----------------------------------------------------------------------------
velox_unpack_fullrows_switch_7_8                            2.78ms    360.19
avx512_new_unpack_fullrows_switch_7_8           384.32%   722.38us     1.38K
----------------------------------------------------------------------------
velox_unpack_fullrows_switch_8_8                            2.78ms    359.51
avx512_new_unpack_fullrows_switch_8_8           514.05%   541.11us     1.85K
----------------------------------------------------------------------------
```

Reply:
I uploaded the codes to this branch https://github.com/yyang52/velox-1/tree/bitpack-cppfiles, you could have a check, thanks!

Reply:
Same from my side. Moving the inline functions from .h to .cpp files would cause great regression. It's said that the modern compiler could automatically do the function inlining, but the performance results did not demonstrate that.

Reply:
Hi @Yuhta, have you checked the performance w/ and w/o inling? Does regression happen on your side? Thanks!

Reply:
Have you added any extra flags on these test files? That makes sense to avoid conflicts with xsimd, but we do not have AVX512 instructions on the `BitPackDecoder.h` anyhow, just function wrappers. The actual implementation is in BitPackDecoderUtilAVX512, does that also affect the build otherwhere?

Reply:
It's compiled with `-O3` as well as a few more options below which I think probably should not matter.  Since `BitPackDecoderUtilAVX512.h` is an header, it will still causes conflicts when we enable `VELOX_ENABLE_AVX512`

```
    "-masm=intel",
    "-mavx",
    "-mavx2",
    "-mbmi2",
    "-mf16c",
    "-mfma",
    "-mlzcnt",
    "-fPIC",
```

Reply:
Here are the flags I dumped, seems like all the needed options are enabled?
```
String dump of section '.GCC.command.line':
  [     0]  /usr/lib/llvm-12/bin/clang --driver-mode=g++ -frecord-command-line -mavx2 -mfma -mavx -mf16c -mlzcnt -std=c++17 -mbmi2 -masm=intel -O3 -Wall -Wextra -Wno-unused -Wno-unused-parameter -Wno-sign-compare -Wno-ignored-qualifiers -Wno-range-loop-analysis -Wno-mismatched-tags -Wno-nullability-completeness -O3 -fPIE -std=gnu++17 -mavx512f -mavx512bw -mavx512vl -mavx512vbmi -mavx512vbmi2 -Werror -D BOOST_ALL_NO_LIB -D BOOST_ATOMIC_DYN_LINK -D BOOST_CONTEXT_DYN_LINK -D BOOST_FILESYSTEM_DYN_LINK -D BOOST_PROGRAM_OPTIONS_DYN_LINK -D BOOST_REGEX_DYN_LINK -D BOOST_SYSTEM_DYN_LINK -D BOOST_THREAD_DYN_LINK -D FMT_LOCALE -D FOLLY_HAVE_INT128_T=1 -D GFLAGS_IS_A_DLL=0 -D VELOX_ENABLE_AVX512 -I _deps/protobuf-src/src -I ../../. -I ../../velox/external/xxhash -I _deps/xsimd-src/include -isystem ../../velox -isystem ../../velox/external -isystem ../../velox/external/duckdb -isystem ../../velox/external/duckdb/tpch/dbgen/include -isystem _deps/gtest-src/googletest/include -isystem _deps/gtest-src/googletest -D USE_VELOX_COMMON_BASE -D HAS_UNCAUGHT_EXCEPTIONS -D NDEBUG -c -o velox/dwio/common/tests/CMakeFiles/velox_dwio_common_bitpack_decoder_benchmarks.dir/BitPackDecoderBenchmarknew.cpp.o ../../velox/dwio/common/tests/BitPackDecoderBenchmarknew.cpp
```